File: goldenflow/transforms/email.py
from __future__ import annotations

import re

import polars as pl

from goldenflow.transforms import register_transform

_EMAIL_PATTERN = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


@register_transform(
    name="email_lowercase",
    input_types=["email", "string"],
    auto_apply=False,
    priority=55,
    mode="series",
)
def email_lowercase(series: pl.Series) -> pl.Series:
    """Lowercase the entire email address."""

    def _lower(val: str | None) -> str | None:
        if val is None:
            return None
        return val.strip().lower()

    return series.map_elements(_lower, return_dtype=pl.Utf8)


@register_transform(
    name="email_normalize",
    input_types=["email"],
    auto_apply=False,
    priority=50,
    mode="series",
)
def email_normalize(series: pl.Series) -> pl.Series:
    """Normalize email: lowercase, strip +tags, strip dots from Gmail local part."""

    def _normalize(val: str | None) -> str | None:
        if val is None:
            return None
        original = val
        val = val.strip().lower()
        if not val or "@" not in val:
            return original  # preserve invalid values
        local, domain = val.rsplit("@", 1)
        # Strip +tag
        local = local.split("+")[0]
        # Strip dots from Gmail local part
        if domain in ("gmail.com", "googlemail.com"):
            local = local.replace(".", "")
        return f"{local}@{domain}"

    return series.map_elements(_normalize, return_dtype=pl.Utf8)


@register_transform(
    name="email_extract_domain",
    input_types=["email"],
    auto_apply=False,
    priority=40,
    mode="series",
)
def email_extract_domain(series: pl.Series) -> pl.Series:
    """Extract the domain from an email address."""

    def _domain(val: str | None) -> str | None:
        if val is None:
            return None
        if "@" not in val:
            return None
        return val.strip().rsplit("@", 1)[1].lower()

    return series.map_elements(_domain, return_dtype=pl.Utf8)


@register_transform(
    name="email_validate",
    input_types=["email", "string"],
    auto_apply=False,
    priority=60,
    mode="series",
)
def email_validate(series: pl.Series) -> pl.Series:
    """Validate email format. Returns True/False/None."""

    def _validate(val: str | None) -> bool | None:
        if val is None:
            return None
        val = val.strip()
        if not val:
            return False
        return bool(_EMAIL_PATTERN.match(val))

    return series.map_elements(_validate, return_dtype=pl.Boolean)

File: packages/goldenflow-js/src/core/transforms/email.ts
/**
 * Email transforms — ported from goldenflow/transforms/email.py
 * Side-effect module: registers 4 email transforms on import.
 */

import type { ColumnValue } from "../types.js";
import { registerTransform } from "./registry.js";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Simple but practical email regex (RFC 5321 simplified). */
const EMAIL_RE = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

/** Gmail-like domains where dots in the local part are ignored. */
const GMAIL_DOMAINS = new Set(["gmail.com", "googlemail.com"]);

// ---------------------------------------------------------------------------
// email_lowercase (55, email|string)
// ---------------------------------------------------------------------------

function emailLowercase(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null || typeof v !== "string") return v;
    return v.toLowerCase();
  });
}

registerTransform(
  { name: "email_lowercase", inputTypes: ["email", "string"], priority: 55, mode: "series" },
  emailLowercase,
);

// ---------------------------------------------------------------------------
// email_normalize (50, email)
// ---------------------------------------------------------------------------

function emailNormalize(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null || typeof v !== "string") return v;
    const lowered = v.toLowerCase().trim();
    const atIdx = lowered.lastIndexOf("@");
    if (atIdx === -1) return lowered;

    let local = lowered.slice(0, atIdx);
    const domain = lowered.slice(atIdx + 1);

    // Strip +tags (e.g. user+tag@gmail.com -> user@gmail.com)
    const plusIdx = local.indexOf("+");
    if (plusIdx !== -1) {
      local = local.slice(0, plusIdx);
    }

    // Strip dots for Gmail-like domains
    if (GMAIL_DOMAINS.has(domain)) {
      local = local.replace(/\./g, "");
    }

    return `${local}@${domain}`;
  });
}

registerTransform(
  { name: "email_normalize", inputTypes: ["email"], priority: 50, mode: "series" },
  emailNormalize,
);

// ---------------------------------------------------------------------------
// email_extract_domain (40, email)
// ---------------------------------------------------------------------------

function emailExtractDomain(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null || typeof v !== "string") return v;
    const atIdx = v.lastIndexOf("@");
    if (atIdx === -1) return null;
    return v.slice(atIdx + 1).toLowerCase();
  });
}

registerTransform(
  { name: "email_extract_domain", inputTypes: ["email"], priority: 40, mode: "series" },
  emailExtractDomain,
);

// ---------------------------------------------------------------------------
// email_validate (60, email|string)
// ---------------------------------------------------------------------------

function emailValidate(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null || typeof v !== "string") return v;
    return EMAIL_RE.test(v.trim());
  });
}

registerTransform(
  { name: "email_validate", inputTypes: ["email", "string"], priority: 60, mode: "series" },
  emailValidate,
);
